Scripts/Runtime/Util/Systems/WorldCache.cs

```diff
@@ -0,0 +1,187 @@
+using System;
+using System.Collections.Generic;
+using Unity.Entities;
+using UnityEngine.LowLevel;
+
+namespace Anvil.Unity.DOTS.Util
+{
+    /// <summary>
+    /// Represents the cached state of a <see cref="World"/>
+    /// </summary>
+    public class WorldCache : AbstractCache
+    {
+        private static readonly Type COMPONENT_SYSTEM_GROUP_TYPE = typeof(ComponentSystemGroup);
+
+        private readonly Dictionary<ComponentSystemBase, SystemCache> m_SystemCacheLookup = new Dictionary<ComponentSystemBase, SystemCache>();
+        private readonly List<SystemCache> m_OrderedSystemCaches = new List<SystemCache>();
+        private readonly List<SystemGroupCache> m_OrderedSystemGroupCaches = new List<SystemGroupCache>();
+
+        private int m_LastRebuildCheckFrameCount;
+        
+        /// <summary>
+        /// The <see cref="World"/> this represents.
+        /// </summary>
+        public World World
+        {
+            get;
+        }
+        
+        internal WorldCache(World world)
+        {
+            World = world;
+        }
+        
+        /// <summary>
+        /// Given a set of <see cref="ComponentType"/>s, populate a list with all systems
+        /// that have <see cref="EntityQuery"/>s that will write/read to any <see cref="ComponentType"/>
+        /// in the set.
+        /// </summary>
+        /// <param name="componentTypes">Set of <see cref="ComponentType"/>s to find matching systems for.</param>
+        /// <param name="matchingSystems">A list that will be populated with the matching systems</param>
+        public void RefreshSystemsWithQueriesFor(HashSet<ComponentType> componentTypes, List<ComponentSystemBase> matchingSystems)
+        {
+            matchingSystems.Clear();
```

Comment:
I'm not though, I'm clearing a lookup `m_OrderedSystemsVersions` instead of the list `m_OrderedSystems` before calling this. 

I think I'll keep this as is for now since we only have one specific use case for it and we can always add an append version later on when someone finds a need. The problem with append is that it's a List and we don't want duplicates. The idea is to get a snapshot of all the systems in order that have queries for these types.